opacus/layers/dp_multihead_attention.py

```diff
@@ -329,3 +332,64 @@ def forward(
             return attn_output, None
+
+    def unsqueeze_0_2(self, t):
+        return torch.unsqueeze(torch.unsqueeze(t, 0), 0)
+
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
+        if destination is None:
+            destination = OrderedDict()
+            destination._metadata = OrderedDict()
+
+        local_metadata = dict(version=self._version)
+        if hasattr(destination, "_metadata"):
+            destination._metadata[prefix[:-1]] = local_metadata
+
+        destination_alter = OrderedDict()
+        if len(prefix.split(".")) > 2:
+            alter_key = ".".join(prefix.split(".")[:-2]) + ".emb.weight"
+        else:
+            alter_key = "emb.weight"
+        if alter_key in destination:
+            destination_alter[alter_key] = destination[alter_key]
+
+        self._save_to_state_dict(destination, prefix, keep_vars)
+        for name, module in self._modules.items():
+            if module is not None:
+                module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
+
+        if self._qkv_same_embed_dim:
+            destination_alter[prefix + 'in_proj_weight'] = torch.cat((
+                destination[prefix + 'qlinear.weight'],
+                destination[prefix + 'klinear.weight'],
+                destination[prefix + 'vlinear.weight'],
+            ), 0)
+        else:
+            destination_alter[prefix + 'q_proj_weight'] = destination[prefix + 'qlinear.weight']
+            destination_alter[prefix + 'k_proj_weight'] = destination[prefix + 'klinear.weight']
+            destination_alter[prefix + 'v_proj_weight'] = destination[prefix + 'vlinear.weight']
+
+        if (prefix + 'qlinear.bias') in destination and (prefix + 'klinear.bias') in destination and (prefix + 'vlinear.bias') in destination:
+            destination_alter[prefix + 'in_proj_bias'] = torch.cat((
+                destination[prefix + 'qlinear.bias'],
+                destination[prefix + 'klinear.bias'],
+                destination[prefix + 'vlinear.bias'],
+            ), 0)
+
+        if self.add_bias_kv:
+            destination_alter[prefix + 'bias_k'] = self.unsqueeze_0_2(destination[prefix + 'seq_bias_k.bias'])
+            destination_alter[prefix + 'bias_v'] = self.unsqueeze_0_2(destination[prefix + 'seq_bias_v.bias'])
+
+        destination_alter[prefix + 'out_proj.weight'] = destination[prefix + 'out_proj.weight']
+        if (prefix + 'out_proj.bias') in destination:
+            destination_alter[prefix + 'out_proj.bias'] = destination[prefix + 'out_proj.bias']
+
+        for hook in self._state_dict_hooks.values():
+            hook_result = hook(self, destination_alter, prefix, local_metadata)
+            if hook_result is not None:
+                destination_alter = hook_result
+
+        return destination_alter
+
+    def named_parameters(self, prefix: str = '', recurse: bool = True):
+        return self.state_dict(prefix = prefix).items()
```

Comment:
Tbh, I would not change `named_parameters` - having this method not match the actual structure of the model would be very confusing and would lead to unexpected behaviour:
* `named_parameters` and `parameters` should return the same set of parameters
* sometimes you need to modify the output of these methods and you want them to point to the actual objects used in the model
* it would mess with calling model attributes directly

Reply:
I think that is a valid concern.
Currently, in `opacus/tests/dp_layers/common.py` line-220, we check if the two `named_parameters` can match.

```
        nn_params = dict(nn_module.named_parameters())
        dp_params = dict(dp_module.named_parameters())
```
How about we just change them into:
```
        nn_params = nn_module.state_dict()
        dp_params = dp_module.state_dict()
```
Since getting `named_parameters` and then change them back to `dict` is a bit weird as well.

Reply:
Yeah, that make sense